File: q4/code.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var findMedianSortedArrays = function(nums1, nums2) {
    var total = nums1.length + nums2.length;
    var isOdd = total & 1;
    var med0, med1;
    if (!isOdd) {
        med1 = (total/2)|0;
        med0 = med1 - 1;
    } else {
        med0 = (total/2)|0;
    }

    var idx1 = 0, idx2 = 0;
    var cnt = 0;
    var sum = 0;
    do {
        var n1 = (nums1[idx1] + 1)|0;
        var n2 = (nums2[idx2] + 1)|0;
        if (n1 == 0 && n2 == 0) { break; }

        if (n1 == 0) {
            if (med0 == cnt) {
                sum += n2;
                if (isOdd) { break; }
            }
            if (med1 == cnt) {
                sum += n2;
                break;
            }
            ++cnt;
            ++idx2;
        } else if (n2 == 0) {
            if (med0 == cnt) {
                sum += n1;
                if (isOdd) { break; }
            }
            if (med1 == cnt) {
                sum += n1;
                break;
            }
            ++cnt;
            ++idx1;
        } else {
            if (n1 < n2) {
                if (med0 == cnt) {
                    sum += n1;
                    if (isOdd) { break; }
                }
                if (med1 == cnt) {
                    sum += n1;
                    break;
                }
                ++cnt;
                ++idx1;
            } else {
                if (med0 == cnt) {
                    sum += n2;
                    if (isOdd) { break; }
                }
                if (med1 == cnt) {
                    sum += n2;
                    break;
                }
                ++cnt;
                ++idx2;
            }
        }
    } while(1);
    return isOdd ? sum - 1 : (sum - 2) / 2;
};

File: q4/code.go
package q4

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	len1 := len(nums1)
	len2 := len(nums2)
	total := len1 + len2

	isOdd := (total & 1) == 1

	var med0 int
	var med1 int
	var l int
	if isOdd {
		med0 = total/2
		med1 = -1
		l = med0
	} else {
		med1 = total/2
		med0 = med1 - 1
		l = med1
	}

	sum := 0
	idx1 := 0
	idx2 := 0
	for i := 0; i < l + 1; i++ {
		n1 := -1
		if idx1 < len1 {
			n1 = nums1[idx1]
		}
		n2 := -1
		if idx2 < len2 {
			n2 = nums2[idx2]
		}

		var n int
		if n1 < 0 {
			idx2++
			n = n2
		} else if n2 < 0 {
			idx1++
			n = n1
		} else if n1 < n2 {
			idx1++
			n = n1
		} else {
			idx2++
			n = n2
		}

		if med0 == i || med1 == i {
			sum += n
		}
	}

	if isOdd {
		return float64(sum)
	} else {
		return float64(sum) / 2
	}
}